File: lib/ruby/cosmos-llm-ruby-tool-preset/lib/cosmos/llm/tool/preset/write.rb
# frozen_string_literal: true

require 'cosmos/llm/tool'

module Cosmos
  module Llm
    module Tool
      module Preset
        # Write file contents preset tool (virtual filesystem)
        #
        # Provides functionality to prepare file writes for a virtual filesystem.
        # Returns the content and metadata needed to update the virtual filesystem.
        #
        # This tool validates the write operation parameters and checks for existing files
        # in the virtual filesystem, then returns structured data containing the write
        # operation details. The actual filesystem modification must be performed by the caller.
        #
        # Key features:
        # - Validates input parameters (file_path and content)
        # - Checks for existing files to determine create vs update operation
        # - Returns comprehensive metadata including size calculations
        # - Handles binary and text content equally
        #
        # @example Writing a new Ruby file
        #   tool = Cosmos::Llm::Tool::Preset.write(filesystem)
        #   result = tool.call(file_path: 'src/new.rb', content: 'puts "hello"')
        #   # result[:success] => true
        #   # result[:created] => true
        #   # result[:size] => 13
        #
        # @example Updating an existing file
        #   tool = Cosmos::Llm::Tool::Preset.write(filesystem)
        #   result = tool.call(file_path: 'existing.txt', content: 'new content')
        #   # result[:success] => true
        #   # result[:updated] => true
        #   # result[:previous_size] => 11 (size of old content)
        #
        # @param filesystem [Cosmos::Llm::VirtualFilesystem::Filesystem] The virtual filesystem to validate against
        # @return [Cosmos::Llm::Tool::Definition] A write file tool definition
        # @raise [ArgumentError] If filesystem parameter is invalid
        # @raise [StandardError] If an error occurs during file lookup or validation
        def self.write(filesystem)
          Cosmos::Llm::Tool.define(:write, register: false) do
            description 'Write content to a file in the virtual filesystem, creating or updating as needed'

            parameter :file_path,
                      type: :string,
                      required: true,
                      description: 'The path to the file to write (relative to virtual filesystem root). Must be a non-empty string.'

            parameter :content,
                      type: :string,
                      required: true,
                      description: 'The content to write to the file. Must be a string (supports both text and binary content).'

            execute do |params|
              file_path = params[:file_path]
              content = params[:content]

              # Input validation
              if file_path.nil? || !file_path.is_a?(String) || file_path.empty?
                raise ArgumentError,
                      'file_path parameter is required and must be a non-empty string'
              end
              raise ArgumentError, 'content parameter must be a string' unless content.is_a?(String)

              begin
                # Check if file already exists in virtual filesystem
                existing_file = filesystem.find_file(file_path)

                {
                  success: true,
                  file_path: file_path,
                  content: content,
                  size: content.bytesize,
                  created: existing_file.nil?,
                  updated: !existing_file.nil?,
                  previous_size: existing_file&.content&.bytesize
                }
              rescue StandardError => e
                {
                  success: false,
                  error: e.message,
                  file_path: file_path
                }
              end
            end
          end
        end
      end
    end
  end
end

# Copyright (c) 2025 Durable Programming, LLC. All rights reserved.

File: lib/rust/cosmos-llm-rust-tool/src/preset/write.rs
use std::sync::Arc;

use cosmos_llm_virtual_filesystem::Filesystem;
use serde_json::json;

use crate::{ParameterType, ToolDefinition};

/// Creates a write-file preset tool bound to `filesystem`.
///
/// The tool validates write parameters and checks whether the file already
/// exists, but does not mutate the filesystem — the caller is responsible
/// for applying the change. Returns metadata describing the operation.
///
/// # Examples
///
/// ```rust
/// use cosmos_llm_virtual_filesystem::Filesystem;
/// use cosmos_llm_tool::preset::write_tool;
/// use serde_json::json;
/// use std::sync::Arc;
///
/// let fs = Arc::new(Filesystem::new("/"));
/// let tool = write_tool(fs);
/// let result = tool.call(&json!({
///     "file_path": "new.txt",
///     "content": "hello",
/// })).unwrap();
/// assert_eq!(result["success"], json!(true));
/// assert_eq!(result["created"], json!(true));
/// ```
pub fn write_tool(filesystem: Arc<Filesystem>) -> ToolDefinition {
    ToolDefinition::new("write")
        .description("Write content to a file in the virtual filesystem")
        .param(
            "file_path",
            ParameterType::String,
            true,
            "Path to the file (relative to virtual filesystem root)",
        )
        .param(
            "content",
            ParameterType::String,
            true,
            "Content to write to the file",
        )
        .handler(move |params| {
            let file_path = params["file_path"]
                .as_str()
                .ok_or("file_path must be a string")?
                .to_string();
            let content = params["content"]
                .as_str()
                .ok_or("content must be a string")?
                .to_string();

            let existing = filesystem.find_file(&file_path);
            let previous_size = existing.and_then(|f| f.content.as_ref()).map(|c| c.len());

            Ok(json!({
                "success": true,
                "file_path": file_path,
                "content": content,
                "size": content.len(),
                "created": existing.is_none(),
                "updated": existing.is_some(),
                "previous_size": previous_size,
            }))
        })
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn write_new_file() {
        let fs = Arc::new(Filesystem::new("/"));
        let t = write_tool(fs);
        let r = t
            .call(&json!({ "file_path": "x.txt", "content": "hi" }))
            .unwrap();
        assert_eq!(r["success"], json!(true));
        assert_eq!(r["created"], json!(true));
        assert_eq!(r["size"], json!(2));
    }

    #[test]
    fn write_existing_file() {
        let fs = Arc::new(Filesystem::build("/", |root| {
            root.file("x.txt", Some("old"), Default::default()).unwrap();
        }));
        let t = write_tool(fs);
        let r = t
            .call(&json!({ "file_path": "x.txt", "content": "new" }))
            .unwrap();
        assert_eq!(r["updated"], json!(true));
        assert_eq!(r["previous_size"], json!(3));
    }
}
